Scale color output in the mobile renderer to provide HDR support

BastiaanOlij opened the pull request:
As we are now using an A2R10G10B10 UNORM buffer our color values can only be between 0.0 and 1.0 disabling full HDR support on the mobile renderer. 

By scaling the color output we can support a 0.0 to 2.0 range. This involves dividing by 2.0 at the output of our scene renderer, and multiplying by 2.0 in our tonemapper. Similar changes are needed in some other parts of the engine such as reading from screen_texture and reading from reflection maps both in reflection probes and for our sky.

Because the clustered renderer and mobile renderer share some code (mostly the lighting code) the `color_scale` input has been added to the clustered `scene_data` as well but remains large unused and should remain at its default `1.0`.

BastiaanOlij committed:
Scale color output in the mobile renderer to provide HDR support

BastiaanOlij commented on servers/rendering/renderer_rd/effects_rd.cpp:
```diff
@@ -59,7 +59,7 @@ RID EffectsRD::_get_uniform_set_from_image(RID p_image) {
 	//any thing with the same configuration (one texture in binding 0 for set 0), is good
-	RID uniform_set = RD::get_singleton()->uniform_set_create(uniforms, luminance_reduce.shader.version_get_shader(luminance_reduce.shader_version, 0), 1);
+	RID uniform_set = RD::get_singleton()->uniform_set_create(uniforms, roughness.shader.version_get_shader(roughness.shader_version, 0), 1);
 
```
So these are the regression fixesses. As the `luminance_reduce` shaders are now unused and not initialised on the mobile renderer, when we abuse them in this way to setup uniform sets for other shaders, stuff break... 

As discussed with Reduz, I plan to split all the effects into their own classes to prevent stuff like this.

BastiaanOlij replied:
This has now been dealt with by rewriting these shaders to fragment shaders in another PR that was already merged.

BastiaanOlij commented on servers/rendering/renderer_rd/effects_rd.cpp:
```diff
@@ -105,7 +105,7 @@ RID EffectsRD::_get_compute_uniform_set_from_texture(RID p_texture, bool p_use_m
 	//any thing with the same configuration (one texture in binding 0 for set 0), is good
-	RID uniform_set = RD::get_singleton()->uniform_set_create(uniforms, luminance_reduce.shader.version_get_shader(luminance_reduce.shader_version, 0), 0);
+	RID uniform_set = RD::get_singleton()->uniform_set_create(uniforms, roughness.shader.version_get_shader(roughness.shader_version, 0), 0);
 
```
Ditto here

BastiaanOlij commented:
This is on hold until I finish #50942 in which I'm also addressing the above breakages in the mobile renderer.

reduz commented:
regarding reading the textures and doing the right adjustment, you can add the code here:
https://github.com/godotengine/godot/blob/master/servers/rendering/renderer_rd/shader_compiler_rd.cpp#L1162
* Add an optional mode (likely a constant that by default is 1.0) called screen_texture_multiply or something you can set.
* Detect that this is reading from a texture (already done in *is_texture_func* )
* Detect that the texture name is SCREEN_TEXTURE ( think you can do this by checking *texture_uniform* ?)
* If all this matches, you need to wrap the whole thing in parenthesis () and after the end where it does `code += ")"`` add something like: `) *  "+rtos(screen_texture_multiply)+")"`

BastiaanOlij commented:
> 
> 
> regarding reading the textures and doing the right adjustment, you can add the code here:
> https://github.com/godotengine/godot/blob/master/servers/rendering/renderer_rd/shader_compiler_rd.cpp#L1162
> 
>     * Add an optional mode (likely a constant that by default is 1.0) called screen_texture_multiply or something you can set.
> 
>     * Detect that this is reading from a texture (already done in _is_texture_func_ )
> 
>     * Detect that the texture name is SCREEN_TEXTURE ( think you can do this by checking _texture_uniform_ ?)
> 
>     * If all this matches, you need to wrap the whole thing in parenthesis () and after the end where it does `code += ")"` add something like: `) *  "+rtos(screen_texture_multiply)+")"`

@reduz Seeing I've already added the scale factor to the scene_data I've just added that in. It was already available as I'm using it in several other places of the shader.
Other then that it seems to be creating the right shader code, I can't test it because reading from screen texture was already broken. Seeing we don't ever seem to make a copy of the render buffer to use as a texture to read to I wonder if this is the bit that is missing here. But that is for another PR methinks.

Let me know if you're happy with the approach so far or if you want it to be changed.

BastiaanOlij commented on servers/rendering/renderer_rd/effects_rd.cpp:
```diff
@@ -903,6 +905,7 @@ void EffectsRD::luminance_reduction_raster(RID p_source_texture, const Size2i p_
 	luminance_reduce_raster.push_constant.min_luminance = p_min_luminance;
 	luminance_reduce_raster.push_constant.exposure_adjust = p_adjust;
+	luminance_reduce_raster.push_constant.color_scale = 2.0; // For our mobile renderer we've halved our color brightness as we're using a UNORM buffer, need to multiply by 2.0
 
```
Forgot about this, this value should be a parameter/setting

reduz commented on servers/rendering/renderer_rd/effects_rd.h:
```diff
@@ -782,7 +782,7 @@ class EffectsRD {
 	void copy_cubemap_to_dp(RID p_source_rd_texture, RID p_dest_texture, const Rect2 &p_rect, float p_z_near, float p_z_far, bool p_dp_flip);
 	void luminance_reduction(RID p_source_texture, const Size2i p_source_size, const Vector<RID> p_reduce, RID p_prev_luminance, float p_min_luminance, float p_max_luminance, float p_adjust, bool p_set = false);
-	void luminance_reduction_raster(RID p_source_texture, const Size2i p_source_size, const Vector<RID> p_reduce, Vector<RID> p_fb, RID p_prev_luminance, float p_min_luminance, float p_max_luminance, float p_adjust, bool p_set = false);
+	void luminance_reduction_raster(RID p_source_texture, const Size2i p_source_size, const Vector<RID> p_reduce, Vector<RID> p_fb, RID p_prev_luminance, float p_min_luminance, float p_max_luminance, float p_adjust, bool p_set = false, float p_color_scale = 2.0);
```
does it make sense to scale the luminance reduction? I would just multiply by scale in the tonemapper, so no precision is lost.

BastiaanOlij replied:
We render into a single channel buffer here so there is alot more precision then the source color. I guess we could do the multiplication much later

BastiaanOlij replied:
I've moved this into the tonemapper, there isn't any difference performance wise but doing this in one place instead of two makes sense.

reduz commented on servers/rendering/renderer_rd/forward_clustered/render_forward_clustered.h:
```diff
@@ -287,6 +287,9 @@ class RenderForwardClustered : public RendererSceneRenderRD {
 
 			uint32_t pancake_shadows;
+
+			float color_scale;
```
This is not really needed in clustered.

BastiaanOlij replied:
The problem is that the clustered and mobile renderer share code, one needs it, the other doesn't...

BastiaanOlij replied:
Now solved with a constant

reduz commented on servers/rendering/renderer_rd/forward_mobile/render_forward_mobile.cpp:
```diff
@@ -271,6 +271,12 @@ bool RenderForwardMobile::free(RID p_rid) {
 
 /* Render functions */
 
+float RenderForwardMobile::_render_buffers_get_color_scale() {
```
I have the feeling this should be called luminance_scale rather than color_scale

reduz replied:
or luminance_multiplier?

BastiaanOlij replied:
I'm open for either

BastiaanOlij replied:
Settled on `luminance_multiplier`

reduz commented on servers/rendering/renderer_rd/shader_compiler_rd.cpp:
```diff
@@ -1197,6 +1198,7 @@ String ShaderCompilerRD::_dump_node_code(const SL::Node *p_node, int p_level, Ge
 							const SL::VariableNode *varnode = static_cast<const SL::VariableNode *>(onode->arguments[i]);
 
 							StringName texture_uniform = varnode->name;
+							is_screen_texture = (texture_uniform == "SCREEN_TEXTURE");
```
All this mechanism most likely needs some enabling in ShaderCompilerRD because, as an example, in 2D or in the clustered rendering, you dont want to use it.

BastiaanOlij replied:
Yeah that makes sense, I'm still learning the parser code so not sure what would be the easiest way to switch on it? Can I check if this is a spatial shader?

BastiaanOlij replied:
@reduz I added a `apply_luminance_multiplier` boolean to `DefaultIdentifierActions` that is only set by the shader compiler for the mobile renderer. Not sure if this is what you had in mind but it seems to do the trick :)

reduz commented on servers/rendering/renderer_rd/shaders/scene_forward_lights_inc.glsl:
```diff
@@ -944,7 +944,7 @@ void reflection_process(uint ref_index, vec3 vertex, vec3 normal, float roughnes
 		vec4 reflection;
 
-		reflection.rgb = textureLod(samplerCubeArray(reflection_atlas, material_samplers[SAMPLER_LINEAR_WITH_MIPMAPS_CLAMP]), vec4(local_ref_vec, reflections.data[ref_index].index), roughness * MAX_ROUGHNESS_LOD).rgb;
+		reflection.rgb = textureLod(samplerCubeArray(reflection_atlas, material_samplers[SAMPLER_LINEAR_WITH_MIPMAPS_CLAMP]), vec4(local_ref_vec, reflections.data[ref_index].index), roughness * MAX_ROUGHNESS_LOD).rgb * scene_data.color_scale;
```
I would probably make this a specialization constant in the mobile renderer and a regular constant (1.0) in the clustered one.

reduz replied:
See #51635 for reference (which I guess should be merged first).

reduz replied:
after merging that PR, on line 400 of shader_forward_mobile.glsl you can add:

`layout(constant_id = 15) const float sc_luminance_multiplier = 2.0;`

This has two advantages. First is that you can also use it to divide at the end of the shader (and it will be turned to a multiplication by the compiler). Second, that in the clustered one you can simply do something like:

`const float sc_luminance_multiplier = 1.0;`

and this will be optimized away.

BastiaanOlij replied:
Ok I'll look into that, need to learn how that works :)
